sdk/servicebus/service-bus/src/core/linkEntity.ts

```diff
@@ -268,4 +310,109 @@ export class LinkEntity {
     }
     return result;
   }
+
+  /**
+   * Determines whether the AMQP link is open. If open then returns true else returns false.
+   * @return {boolean} boolean
+   */
+  isOpen(): boolean {
+    const result: boolean = this._link ? this._link.isOpen() : false;
+    log.error(`${this._logPrefix} is open? ${result}`);
+    return result;
+  }
+
+  /**
+   * Indicates that _closeLink("close") has been called on this link and
+   * that it should not be allowed to reopen.
+   */
+  private _wasClosedByUser: boolean = false;
+
+  protected get wasClosedByUser(): boolean {
+    return this._wasClosedByUser;
+  }
+
+  /**
+   * Indicates that a link initialization is in process.
+   */
+  public get isConnecting(): boolean {
+    return this._isConnecting;
+  }
+
+  private _isConnecting: boolean = false;
+
+  /**
+   * NOTE: This method should be implemented by any child classes to actually create the underlying
+   * Rhea link (AwaitableSender vs Receiver vs RequestResponseLink)
+   *
+   * @param _options
+   */
+  protected abstract createRheaLink(_options: LinkOptionsT<LinkT>): Promise<LinkT>;
+
+  /**
+   * Initializes this LinkEntity, setting this._link with the result of  `createRheaLink`, which
+   * is implemented by child classes.
+   *
+   * @returns A Promise that resolves when the link has been properly initialized
+   */
+  async initLink(options: LinkOptionsT<LinkT>, abortSignal?: AbortSignalLike): Promise<void> {
+    const checkAborted = (): void => {
+      if (abortSignal?.aborted) {
+        throw new AbortError(StandardAbortMessage);
+      }
+    };
+
+    const connectionId = this._context.namespace.connectionId;
+    checkAborted();
+
+    if (options.name) {
+      this.name = options.name;
+      this._logPrefix = `[${connectionId}|l:${this.name}|a:${this.address}]`;
+    }
+
+    if (this._wasClosedByUser) {
+      log.error(`${this._logPrefix} Link has been closed by user. Not reopening.`);
+      return;
+    }
+
+    if (this.isOpen()) {
+      log.error(`${this._logPrefix} Link is already open. Returning.`);
+      return;
+    }
+
+    if (this.isConnecting) {
+      log.error(`${this._logPrefix} Link is currently opening. Returning.`);
+      return;
+    }
+
+    log.error(`${this._logPrefix} Is not open and is not currently connecting. Opening.`);
+
+    this._isConnecting = true;
+
+    try {
+      await this._negotiateClaim();
+      checkAborted();
+
+      log.error(`${this._logPrefix} Creating with options %O`, options);
+      this._link = await this.createRheaLink(options);
+      checkAborted();
+
+      if (this._wasClosedByUser) {
+        // the user attempted to close while we were still initializing the link. Go
+        // ahead and close the link we created.
+        log.error(`${this._logPrefix} User closed link while it was initializing.`);
+        throw new AbortError("Link closed by user while initializing.");
```

Comment:
Are we sure we hit this `if` block? Am sure you have tests, but am thinking of the below:
- `_wasClosedByUser` is set to true when `closeLink()` is called.
- `closeLink()` is gated with `if (this.link)`
- So, when the first time `initLink()` is called and then `closeLink()` gets called before the former completes, then `_wasClosedByUser` would not be set to true. 

Also, why throw AbortError instead of closing the `link` as the comment says?

Reply:
Renamed the flag to just be "_wasClosed" which I think is generic enough to cover all the cases.

The AbortError one is an interesting point. I've been thinking more about what we have as built-in error types that make sense. AbortError has two nice properties for this - it's non-retryable (as a force close would be) and from a name perspective it matches what's happening here.

Just closing it might make sense but I thought it'd also be good to basically say "init() succeeds which means _link is defined _or_ it throws" rather than having some cases where init() doesn't throw but _link is not set

. The only spot that doesn't hold (and it might be nice to fix) is if the link is _already_ opening but I haven't thought about how to appropriately plug that yet. Perhaps doing the 'lock wait' pattern we have for some things so at least it blocks until one of the initializations completes.

Reply:
Okay, the sequence that I've got encoded in the test (which I think makes sense) is:

1. Link is not open
2. User calls initLink()
3. Just after negotiateClaim (but before createRheaLink) the user initiates a closeLink()
4. Now the user has logically asked us to close the link but we're createRheaLink is already scheduled
5. _wasClosed check fires and properly closes down the link it just created.

This replaces some code I had before where we would chain the init() promise (in subscribe, I believe) with a check to see if wasClosed had been called. I figured it was better to just make it part of the general logic of init() rather than requiring that everywhere else (became more obvious once I made _wasClosed a private field of the base class).

(this is the test, btw)
https://github.com/Azure/azure-sdk-for-js/blob/b475adb77c699d2523979aa63d4995b2d8d8646a/sdk/servicebus/service-bus/test/internal/linkentity.unittest.spec.ts#L196

Reply:
In step 3 above, you have the user initiating a `close()` on the client/sender/receiver exposed to them which then calls `close()` on the `MessageSender`/`MessageReceiver`/`ManagementClient` instances. In some of these classes, we then call `closeLink()` directly and what you say above holds true. But in some of these classes we check  for `if(this.link)` before calling `closeLink()`. I was referring to these cases.

In such cases `if(this.link)`check will fail and the flag will not be set.

Taking a step back, from what I understand, the need to track the user initiated close is required only to stop the background process of recovering broken links.Since the `LinkEntity` is not responsible for the link recovery process, perhaps it should not care about tracking this flag at all and the `StreamingReceiver` and `MessageSender` should instead track this and make use of the abort signal to cancel any ongoing init() process?

Reply:
> Just closing it might make sense but I thought it'd also be good to basically say "init() succeeds which means _link is defined or it throws" rather than having some cases where init() doesn't throw but _link is not set

In `initLink()`, we do check for `_wasClosedPermanently` before starting any async work. If true, we simply return. Wont this result in `init()` succeeding but no `_link`?

Reply:
Looks like we have removed the `if(this.link)` check before calling `closeLink()`, so most of my concerns above are resolved. The one right above this comment is the only one pending

Reply:
> In initLink(), we do check for _wasClosedPermanently before starting any async work. If true, we simply return. Wont this result in init() succeeding but no _link?

Absolutely - it's a hole I'd like to patch. I mention it here: https://github.com/Azure/azure-sdk-for-js/issues/10656

Reply:
(resolving this in favor of the issue I submitted)